File: leetcode/lc-us/cpp/1824.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int minSideJumps(vector<int> &obstacles) {
        const int INF = 1e9;
        int a = 1, b = 0, c = 1;
        for (int i = 1; i < obstacles.size(); i++) {
            int x = INF, y = INF, z = INF;
            switch (obstacles[i]) {
                case 1:
                    y = min(b, c + 1);
                    z = min(c, b + 1);
                    break;
                case 2:
                    x = min(a, c + 1);
                    z = min(c, a + 1);
                    break;
                case 3:
                    x = min(a, b + 1);
                    y = min(b, a + 1);
                    break;
                default:
                    x = min(a, min(b, c) + 1);
                    y = min(b, min(a, c) + 1);
                    z = min(c, min(a, b) + 1);
                    break;
            }
            a = x, b = y, c = z;
        }
        return min(a, min(b, c));
    }
};

File: leetcode/lc-us/py/1824.py
from typing import List


class Solution:
    def minSideJumps(self, obstacles: List[int]) -> int:
        a, b, c = 1, 0, 1
        for i in range(1, len(obstacles)):
            o = obstacles[i]
            a1 = b1 = c1 = float('inf')
            if o == 1:
                b1, c1 = min(b, c + 1), min(c, b + 1)
            elif o == 2:
                a1, c1 = min(a, c + 1), min(c, a + 1)
            elif o == 3:
                a1, b1 = min(a, b + 1), min(b, a + 1)
            else:
                a1, b1, c1 = min(a, b + 1, c + 1), min(b, a + 1, c + 1), min(c, a + 1, b + 1)
            a, b, c = a1, b1, c1
        return min(a, b, c)
